source/_posts/2018/2018-04-23-how-to-slowly-turn-your-symfony-project-to-legacy-with-action-injection.md

```diff
@@ -0,0 +1,366 @@
+```
+
+### What are Propagated Advantages?
+
+- less writing
+- manual wiring of only used services - with no benchmark this has similar value as statements like "Symfony is 3x faster than Laravel, it's true"
+- smaller controllers
+
+### What are Already Known Disadvantages?
+
+Paul M. Jones wrote that [“Action Injection” As A Code Smell](http://paul-m-jones.com/archives/6589). Why?
+
+<blockquote class="blockquote">
+"The fact that your controller has so many dependencies, used only in some cases and not in others, <strong>should be an indicator that the class is doing too much</strong>. Indeed, it’s doing so much that you cannot call its action methods directly; you have to use the dependency injection container not only to build the controller object but also to invoke its action methods."
+</blockquote>
+
+And I agree. It's the same code smell as adding 10th action method to the `ProductController` that now has 300 lines. Maybe you should split it into 2 classes and add [sniff](https://github.com/object-calisthenics/phpcs-calisthenics-rules#7-keep-your-classes-small) to make sure this won't happen in production code ever again (because no-one else will do it better than continuous integration).
+
+But that's just words and ideas, no legacy (yet).
+
+What might really happen with *autowired arguments* approach?
+
+## 1. Injection Everywhere
+
+<img src="/assets/images/posts/2018/action-injection/everywhere.jpg" class="img-thumbnail">
+
+## The Nette-Framework-Tried-It-For-You story
+
+Nette "inject" feature released in 2014 in Nette 2.1 started very similarly. It has 2 ways to inject dependencies:
+
+### A. `@inject` Annotatoin
+
+```php
+namespace App\Controller;
+
+use App\Model\SomeService;
+
+final class SomeController
+{
+    /**
+     * @var SomeService
+     * @inject
+     */
+    public $someService;
+}
+```
+
+### B. `inject*()` Method
+
+```php
+namespace App\Controller;
+
+use App\Model\SomeService;
+
+final class SomeController
+{
+    /**
+     * @var SomeService
+     */
+    private $someService;
+
+    public function injectSomeService(SomeService $someService)
+    {
+        $this->someService = $someService;
+    }
+}
+```
+
+It also have to be activated in config manually with specific `tags: ['inject']` tag, as in Symfony:
+
+```yaml
+# app/config/services.neon
+services:
+    App\Controller\SomeController:
+        tags: ['inject']
+
+    - App\Model\SomeService
+```
+
+Can you see the difference to Symfony? Well, almost none. But so far so good.
+
+Note to Nette programmers: [`@inject` is often a code smell and you should do it cleaner](/blog/2016/12/24/how-to-avoid-inject-thanks-to-decorator-feature-in-nette/)
+
+## Inspire by (Good/Bad) Example
+
+If you prepare some "dirty-hack-that-none-should-use" or even better "don't-ever-use-this-unless-you-know-why" and make it public, you can be sure people will ignore it and use it a very creative way. Unless there is `new ForbiddenUseException` thrown.
+
```

Comment:
Shouldn't it be "use it in a very creative way"?

Reply:
You're right of course. I'll fix it

How do you understand post as a whole?

Reply:
Don't overuse action injection?

Reply:
Better not use it at all

Reply:
How would you inject Request to controller method? Using constructor RequestStack injection?

Reply:
Yep. Single approach